server/src/main/java/org/elasticsearch/index/engine/SearchBasedChangesSnapshot.java

```diff
@@ -0,0 +1,233 @@
+
+    private final IndexVersion indexVersionCreated;
+    private final IndexSearcher indexSearcher;
+    private final Closeable onClose;
+
+    protected final long fromSeqNo, toSeqNo;
+    protected final boolean requiredFullRange;
+    protected final int searchBatchSize;
+
+    private final boolean accessStats;
+    private final int totalHits;
+    private FieldDoc afterDoc;
+    private long lastSeenSeqNo;
+
+    /**
+     * Constructs a new snapshot for fetching changes within a sequence number range.
+     *
+     * @param engineSearcher       Engine searcher instance.
+     * @param searchBatchSize      Number of documents to retrieve per batch.
+     * @param fromSeqNo            Starting sequence number.
+     * @param toSeqNo              Ending sequence number.
+     * @param requiredFullRange    Whether the full range is required.
+     * @param accessStats          If true, enable access statistics for counting total operations.
+     * @param indexVersionCreated  Version of the index when it was created.
+     */
+    protected SearchBasedChangesSnapshot(
+        Engine.Searcher engineSearcher,
+        int searchBatchSize,
+        long fromSeqNo,
+        long toSeqNo,
+        boolean requiredFullRange,
+        boolean accessStats,
+        IndexVersion indexVersionCreated
+    ) throws IOException {
+
+        if (fromSeqNo < 0 || toSeqNo < 0 || fromSeqNo > toSeqNo) {
+            throw new IllegalArgumentException("Invalid range; from_seqno [" + fromSeqNo + "], to_seqno [" + toSeqNo + "]");
+        }
+        if (searchBatchSize <= 0) {
+            throw new IllegalArgumentException("Search_batch_size must be positive [" + searchBatchSize + "]");
+        }
+
+        this.indexVersionCreated = indexVersionCreated;
+        this.fromSeqNo = fromSeqNo;
+        this.toSeqNo = toSeqNo;
+        this.lastSeenSeqNo = fromSeqNo - 1;
+        this.requiredFullRange = requiredFullRange;
+        this.indexSearcher = newIndexSearcher(engineSearcher);
+        this.indexSearcher.setQueryCache(null);
+
+        final AtomicBoolean closed = new AtomicBoolean();
+        this.onClose = () -> {
+            if (closed.compareAndSet(false, true)) {
+                IOUtils.close(engineSearcher);
+            }
+        };
+
+        long requestingSize = (toSeqNo - fromSeqNo == Long.MAX_VALUE) ? Long.MAX_VALUE : (toSeqNo - fromSeqNo + 1L);
+        this.searchBatchSize = (int) Math.min(requestingSize, searchBatchSize);
+
+        this.accessStats = accessStats;
+        this.totalHits = accessStats ? indexSearcher.count(rangeQuery(fromSeqNo, toSeqNo, indexVersionCreated)) : -1;
+    }
+
+    /**
+     * Abstract method for retrieving the next operation. Should be implemented by subclasses.
+     *
+     * @return The next Translog.Operation in the snapshot.
+     * @throws IOException If an I/O error occurs.
+     */
+    public abstract Translog.Operation nextOperation() throws IOException;
```

Comment:
nit: this can be protected instead.

Reply:
Updated in https://github.com/elastic/elasticsearch/pull/114618/commits/f7024efdda0362164194621ed074f49dd4844eae